From 06157e70b86f124a48d729815f6da1639fe2a06c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Bastien=20ROUCARI=C3=88S?= <roucaries.bastien@gmail.com>
Date: Wed, 26 Jul 2017 18:29:54 +0200
Subject: [PATCH] Fix a memory  leak in enhance.c

Fix a potential memory leak if memory could not be allocated for one
of histogram or stretch_map.
If both cannot be allocated, there is no memory leak. If only one is
allocated and the other fails,
there is a memory leak of the one that could not be allocated. There
is very little chance the allocations would fail, so its low risk.

bug: https://github.com/ImageMagick/ImageMagick/commit/0b13c96bbb4d7ca9b738670524665fa13e5d0dab
bug-debian: https://bugs.debian.org/869769
origin: https://github.com/ImageMagick/ImageMagick/commit/0b13c96bbb4d7ca9b738670524665fa13e5d0dab
---
 magick/enhance.c | 10 ++++++++--
 1 file changed, 8 insertions(+), 2 deletions(-)

diff --git a/magick/enhance.c b/magick/enhance.c
index d4d91b76a..26308b67f 100644
--- a/magick/enhance.c
+++ b/magick/enhance.c
@@ -1178,8 +1178,14 @@ MagickExport MagickBooleanType ContrastStretchImageChannel(Image *image,
     sizeof(*stretch_map));
   if ((histogram == (MagickPixelPacket *) NULL) ||
       (stretch_map == (QuantumPixelPacket *) NULL))
-    ThrowBinaryException(ResourceLimitError,"MemoryAllocationFailed",
-      image->filename);
+    {
+      if (stretch_map != (double *) NULL)
+        stretch_map=(double *) RelinquishMagickMemory(stretch_map);
+      if (histogram != (double *) NULL)
+        histogram=(double *) RelinquishMagickMemory(histogram);
+      ThrowBinaryException(ResourceLimitError,"MemoryAllocationFailed",
+        image->filename);
+    }
   /*
     Form histogram.
   */
